nucliadb_node/src/replication/replicator.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (C) 2021 Bosutech XXI S.L.
+//
+// nucliadb is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at info@nuclia.com.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fs;
+use std::path::Path;
+use std::sync::Arc;
+
+use nucliadb_core::metrics::metric::replication as replication_metrics;
+use nucliadb_core::tracing::{error, info, warn};
+use nucliadb_core::{metrics, Error, NodeResult};
+use nucliadb_protos::replication;
+use tokio::io::AsyncWriteExt;
+use tonic::Request;
+
+use crate::replication::health::ReplicationHealthManager;
+use crate::settings::Settings;
+use crate::shards::metadata::ShardMetadata;
+use crate::shards::providers::unbounded_cache::AsyncUnboundedShardWriterCache;
+use crate::shards::providers::AsyncShardWriterProvider;
+use crate::shards::writer::ShardWriter;
+use crate::utils::list_shards;
+
+pub async fn replicate_shard(
```

Comment:
I think this function could be simplified and some allocations may be removed if instead of using `String` for paths we use `std::path::PathBuf`.

Reply:
good point! This is silly.